### PR TITLE
Reintroduce Set Similarity under Jaccard Distance

### PR DESCRIPTION
**Pull request by maumueller**
This PR reintroduces support for ANN for sets under Jaccard similarity. (It always bothered me that we lost support for a 1k star ANN implementation https://github.com/ekzhu/datasketch/.) Currently, there is only one standard data set and some random sets for testing.

I tried to keep the PR as minimally invasive as possible. This means that sets are stored as bit vector representation in hdf5 (because they are supposed to be 2d np matrices), but are then transformed to a sparse list-of-indices representation (needed by the algorithms).

Looking forward to hearing your thoughts!

### CODE REVIEW

**Comment by maumueller** [comment]
The current leader board is a bit empty, so it might not be worth adding it to the README :-)

kosarak-jaccard

**Comment by erikbern** [comment]
nice! i think jaccard could probably be added to annoy in some hacky way.

File: ann_benchmarks/datasets.py [hunk 1]
```diff
@@ -279,6 +283,50 @@ def sift_hamming(out_fn, fn):
...
+    for line in content:
+        if len(line.split()) >= min_elements:
+            for x in line.split():
+                X[i][ids[int(x)]] = 1
+            i += 1
+
+    X_train, X_test = train_test_split(numpy.array(X), test_size=500)
+    write_output(X_train, X_test, out_fn, 'jaccard', 'bit')
+
+def random_jaccard(out_fn, n=10000, size=50, universe=80):
```
Comment (erikbern) [inline ann_benchmarks/datasets.py]: can you seed the random here just so it's reproducible?

File: ann_benchmarks/distance.py [hunk 2]
```diff
@@ -1,20 +1,38 @@
...
-    intersect = len(a & b)
+         return 0
+    intersect = len(set(a) & set(b))
     return intersect / (float)(len(a) + len(b) - intersect)
 
+def transform_dense_to_sparse(X):
+    """Converts the n * m dataset into a sparse format
+    that only holds the non-zero entries (Jaccard)."""
+    # get list of indices of non-zero elements
+    indices = np.transpose(np.where(X))
```
Comment (erikbern) [inline ann_benchmarks/distance.py]: not super important but you could probably use `itertools.groupby` here:
```python
indices = np.transpose(np.where(X))
keys = []
for ijs in itertools.groupby(indices,  lambda ij: ij[0]):
    keys.append([j for i, j in ijs])
return keys
```
haven't tested it though

File: test/test-jaccard.py [hunk 3]
```diff
@@ -0,0 +1,22 @@
+import unittest
```
Comment (erikbern) [inline test/test-jaccard.py]: thanks for the unit test!

File: ann_benchmarks/distance.py (hunk 2 above)
Comment (maumueller) [reply ann_benchmarks/distance.py]: Thanks, worked almost out-of-the-box. :-)

File: ann_benchmarks/datasets.py (hunk 1 above)
Comment (maumueller) [reply ann_benchmarks/datasets.py]: Sure! Added. (There are multipe `random_state=1` and `seed(1)` calls now. Maybe there should be a unique global seed.)

**Comment by maumueller** [comment]
Hi Erik. What is the status on this? Is there anything you would like to discuss/see before merging?

**Comment by erikbern** [comment]
sorry @maumueller i just forgot about it. feel free to ping me quicker next time in case a few days pass!
